lib/io.js

```diff
@@ -51,22 +51,31 @@ BinaryDecoder.prototype = {
         var oldOffset = this._input.offset;
         var b = this.readByte();
         if (b instanceof AvroErrors.BlockDelayReadError) return b;
+        b = Long.fromBits(b, 0);
 
-        var n = b & 0x7F;
+        var L_0x7f = Long.fromNumber(0x7F);
+        var L_0xFF = Long.fromNumber(0xFF);
+
+        var n = b.and(L_0x7f);
         var shift = 7;
+        var numShifts = 0;
 
-        while ((b & 0x80) != 0) {
+        while (b.greaterThan(L_0x7f) && numShifts < 7) {
             b = this.readByte();
             if (b instanceof AvroErrors.BlockDelayReadError) {
                 //console.error("offset now is %d, old was %d, shift is %d", this._input.offset, oldOffset, shift);
                 this._input.skip(oldOffset - this._input.offset);
                 return b;
             }
-            n |= (b & 0x7F) << shift
```

Comment:
Wouldn't just changing this to `n ^= (b & 0x7F) << shift` fix this? You'd still need to track the `numShifts` and use it on the `while (b > 0x7F && numShifts < 7)`, but it seems that it would take care of this and avoid using the `Long` module (which is pretty awesome, by the way, I can use this elsewhere)